Backend doesn't gracefully quit if sub-components fail to start

E.g. dashboard was not able to start due to port 3000 already being in use. I got the following exception:

```go
{"component":"dashboard","level":"info","msg":"starting dashboardd on address: 0.0.0.0:3000","time":"2017-05-16T12:14:35-07:00"}
panic: sync: negative WaitGroup counter

goroutine 221 [running]:
sync.(*WaitGroup).Add(0xc42034cd90, 0xffffffffffffffff)
        /home/justin/go/src/sync/waitgroup.go:75 +0x1c9
sync.(*WaitGroup).Done(0xc42034cd90)
        /home/justin/go/src/sync/waitgroup.go:100 +0x34
github.com/sensu/sensu-go/backend/dashboardd.(*Dashboardd).Start.func1(0xc4206d0690, 0xc4206be160)
        /home/justin/code/go/src/github.com/sensu/sensu-go/backend/dashboardd/dashboardd.go:70 +0x56
created by github.com/sensu/sensu-go/backend/dashboardd.(*Dashboardd).Start
        /home/justin/code/go/src/github.com/sensu/sensu-go/backend/dashboardd/dashboardd.go:70 +0x385
2017-05-16 12:14:36.098784 I | backend not ready, sleeping...
```

--- a/backend/apid/apid.go
+++ b/backend/apid/apid.go
@@ -17,10 +17,11 @@ import (
 
 // APId is the backend HTTP API.
 type APId struct {
-	stopping chan struct{}
-	running  *atomic.Value
-	wg       *sync.WaitGroup
-	errChan  chan error
+	stopping   chan struct{}
+	running    *atomic.Value
+	wg         *sync.WaitGroup
+	errChan    chan error
+	httpServer *http.Server
 
 	Authentication authentication.Provider
 	BackendStatus  func() types.StatusMap
@@ -44,18 +45,23 @@ func (a *APId) Start() error {
 	router := httpRouter(a)
 	routerStack := authentication.Middleware(a.Authentication, router)
 
-	server := &http.Server{
+	a.httpServer = &http.Server{
 		Addr:         fmt.Sprintf("%s:%d", a.Host, a.Port),
 		Handler:      routerStack,
 		WriteTimeout: 15 * time.Second,
 		ReadTimeout:  15 * time.Second,
 	}
 
-	logger.Info("starting apid on address: ", server.Addr)
+	logger.Info("starting apid on address: ", a.httpServer.Addr)
+	a.wg.Add(1)
 
 	go func() {
 		defer a.wg.Done()
-		server.ListenAndServe()
+		if err := a.httpServer.ListenAndServe(); err != nil {
+			// TODO (JK): need a way to handle closing things like errChan, etc.
+			// in cases where there's a failure to start the daemon
+			logger.Errorf("failed to start http server: %s", err.Error())
+		}
 	}()
 
 	return nil
@@ -63,6 +69,14 @@ func (a *APId) Start() error {
 
 // Stop httpApi.
 func (a *APId) Stop() error {
+	if err := a.httpServer.Shutdown(nil); err != nil {
+		// failure/timeout shutting down the server gracefully
+		logger.Error("failed to shutdown http server gracefully - forcing shutdown")
+		if closeErr := a.httpServer.Close(); closeErr != nil {
+			logger.Error("failed to shutdown http server forcefully")
+		}
+	}
+
 	a.running.Store(false)
 	close(a.stopping)
 	a.wg.Wait()

--- a/backend/dashboardd/dashboardd.go
+++ b/backend/dashboardd/dashboardd.go
@@ -33,6 +33,7 @@ type Dashboardd struct {
 	wg            *sync.WaitGroup
 	errChan       chan error
 	BackendStatus func() types.StatusMap
+	httpServer    *http.Server
 
 	Config
 }
@@ -55,18 +56,21 @@ func (d *Dashboardd) Start() error {
 
 	router := httpRouter(d)
 
-	server := &http.Server{
+	d.httpServer = &http.Server{
 		Addr:         fmt.Sprintf("%s:%d", d.Host, d.Port),
 		Handler:      router,
 		WriteTimeout: 15 * time.Second,
 		ReadTimeout:  15 * time.Second,
 	}
 
-	logger.Info("starting dashboardd on address: ", server.Addr)
+	logger.Info("starting dashboardd on address: ", d.httpServer.Addr)
+	d.wg.Add(1)
 
 	go func() {
 		defer d.wg.Done()
-		server.ListenAndServe()
+		if err := d.httpServer.ListenAndServe(); err != nil {
+			logger.Errorf("failed to start http server: %s", err.Error())
+		}
 	}()
 
 	return nil
@@ -74,6 +78,14 @@ func (d *Dashboardd) Start() error {
 
 // Stop dashboardd.
 func (d *Dashboardd) Stop() error {
+	if err := d.httpServer.Shutdown(nil); err != nil {
+		// failure/timeout shutting down the server gracefully
+		logger.Error("failed to shutdown http server gracefully - forcing shutdown")
+		if closeErr := d.httpServer.Close(); closeErr != nil {
+			logger.Error("failed to shutdown http server forcefully")
+		}
+	}
+
 	close(d.stopping)
 	d.wg.Wait()
 	close(d.errChan)

